gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,149 @@
+// Copyright (C) 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+)
+
+// TransformChain is the wrapper to run transforms
+type transformChain struct {
+	transforms []Transform
+	out        Writer
+}
+
+func CreateTransformChain(out Writer, transforms []Transform) *transformChain {
+	return &transformChain{
+		transforms: transforms,
+		out:        out,
+	}
+}
+
+func (chain *transformChain) BeginChain(ctx context.Context) error {
+	handleInitialState(chain.out.State())
+
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.BeginTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+		}
+
+		if transform.RequiresAccurateState() {
+			err = mutateAndWrite(ctx, 0, cmds, chain.out)
+			if err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	err = mutateAndWrite(ctx, 0, cmds, chain.out)
+	if err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) EndChain(ctx context.Context) error {
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.EndTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "End Transform Error [%v] : %v", transform, err)
+		}
+
+		if transform.RequiresAccurateState() {
+			err = mutateAndWrite(ctx, 0, cmds, chain.out)
+			if err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	err = mutateAndWrite(ctx, 0, cmds, chain.out)
+	if err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
```

Comment:
At a glance these few functions seem to be almost entirely identical. May want to refactor somewhat.